controller/lib/hal/hal.h

```diff
@@ -238,6 +238,10 @@ class HalApi {
   void InitPSOL();
   void PSOL_Value(float val);
 
+  // Flash access
+  bool FlashErasePage(uint32_t address);
+  bool FlashWrite(uint32_t addr, uint32_t *data, int ct);
```

Comment:
nvparams will compile for the native platform, because all of the code that touches the STM32 registers is hidden under HAL.  But it doesn't actually *work* on native, because it reads from a fixed address and expects that the flash data will be present there.  IOW the nominally platform-independent nvparams.cpp code is actually not platform-independent.

I can see a few ways to fix this, but fundamentally I think these functions in HAL are probably at the wrong level of abstraction.  Instead to me it seems like HAL should expose something like

```
struct NVParams {
  uint32_t calibration_foo;
  float calibration_bar;
};
const NVParams& GetNVParams();
void SetNVParams(const NVParams& params);
```

(Now it's trivial to write a meaningful native/fake implementation of this!)

Then nvparams.cpp gets renamed to stm32_flash and has something like

```
// OK, I'm overloading the meaning of "flash page" -- we should probably have a different name.
struct FlashPageHeader {
  uint32_t crc;
  // other header stuff
};
struct FlashPage {
  FlashPageHeader header;
  NVParams params;
  char rsvd[512 - sizeof(header) - sizeof(params)];
};

// As written this ensures there's no padding between FlashPageHeader and NVParams, which is important!
static_assert(sizeof(FlashPage) == 512));
```

I realize this is a big structural change to the code, and I understand if you're not willing to do it.  I don't think this is a stylistic thing, because afaict fundamentally the code as-is *can't* work natively.  If you can come up with a way for it to work with a smaller change (or perhaps without any changes at all), I can make this change on my own time.  Or you could even remove some code so that I'm able to do this without having to *change* existing code.

The need to change things also brings up for me the need to be able to test things.  I would really like to have some stm32 tests running for this, again just because we might need to make some changes.  For instance, I would like to replace the opaque stm32 registers with the bitfield regs, but I don't feel particularly confident making that change without tests.

I don't block this review on tests, but if you're not willing to do that, perhaps you'd leave a TODO or file a bug.

Reply:
(Actually this ensures there's no external padding between FlashPageHeader and NVParams, but doesn't ensure there's no padding internally inside FlashPageHeader, which is also important.  Anyway we can figure it out.)

Reply:
We could add a simulated flash interface for native testing which reads/writes to a file or something.  When you say it wont work under native I'm not sure what you mean.  Nothing really works under native because there's no hardware to back it up.  The native environment is really just for testing, right?
You're certainly welcome to rewrite this to suit your taste, I won't be offended.  I just wanted to get some non-volatile parameter support implemented so we could move forward with saving calibration data.

Reply:
> When you say it wont work under native I'm not sure what you mean

As currently written, sensors.cpp is going to do something like

```
uint32_t dp_zero_point = FindStore()->dp_zero_point;
```

We want this not to crash when we run the `sensors_test.cpp` unit test on x86.

As written, `FindStore` will crash unless we first initialize the nvparams module.  When we run `NVparamsInit()`, it reads from the hardcoded flash memory address.  Because that address is not mapped on x86, it will crash.

Reply:
> We could add a simulated flash interface for native testing which reads/writes to a file or something.

Yeah, that's basically what I'm saying we need, otherwise we cannot unit test any module which relies on non-volatile params.  My suggestion was to write this simulated interface at a higher level -- instead of it being a simulated flash memory interface, it could be a simulated store of non-volatile params.  That's what I was trying to mock out above.

The reason I was suggesting it be a higher-level interface is because it's hard to simulate the low-level flash API.  For instance, as written we'd have to map a file to a particular address that may or may not be free on x86.  It seems to me that it would be better to mock one layer above raw flash pages.

Reply:
Yup, I see what you mean and I agree that it would be difficult to write a native implementation of the low level flash parameters.  I'll move the NVparams code into the HAL since it's so hardware specific.